### PR TITLE
Add symbolic call graph for `Product` block encoding

### PR DESCRIPTION
**Pull request by charlesyuan314**
See https://github.com/quantumlib/Qualtran/issues/1309#issuecomment-2313695831.

### CODE REVIEW

File: qualtran/bloqs/block_encoding/product.py [hunk 1]
```diff
@@ -132,6 +135,61 @@ def epsilon(self) -> SymbolicFloat:
...
+                    regs.append(Unused(self.resource_bitsize - u.resource_bitsize))
+                partition.append((Register("resource", dtype=QAny(u.resource_bitsize)), regs))
+            ret.append(AutoPartition(u, partition, left_only=False))
+        return ret
+
+    def build_call_graph(self, ssa: SympySymbolAllocator) -> Set[BloqCountT]:
+        counts = Counter[Bloq]()
+        if is_symbolic(self.ancilla_bitsize) or self.ancilla_bitsize > 0:
+            counts[self.anc_part] += 1
+            counts[evolve(self.anc_part, partition=False)] += 1
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/product.py:179]: I think we usually don't count split/join/partition bloqs in `build_call_graph`? @mpharrigan, thoughts?

File: qualtran/bloqs/block_encoding/product.py [hunk 2]
```diff
@@ -132,6 +135,61 @@ def epsilon(self) -> SymbolicFloat:
...
+        for bloq in self.constituents:
+            counts[bloq] += 1
+        n = len(self.block_encodings)
+        for i, u in enumerate(reversed(self.block_encodings)):
+            if (
+                (is_symbolic(u.ancilla_bitsize) or u.ancilla_bitsize > 0)
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/product.py:185]: you can use the new `is_zero` function in `symbolics`

File: qualtran/bloqs/block_encoding/product.py [hunk 3]
```diff
@@ -132,6 +135,61 @@ def epsilon(self) -> SymbolicFloat:
...
+    @property
+    def constituents(self) -> Sequence[Bloq]:
+        n = len(self.block_encodings)
+        anc_bits = self.ancilla_bitsize - (n - 1)
+        ret = []
+        for u in reversed(self.block_encodings):
+            partition: List[Tuple[Register, List[Union[str, Unused]]]] = [
+                (Register("system", dtype=QAny(u.system_bitsize)), ["system"])
+            ]
+            if is_symbolic(u.ancilla_bitsize) or u.ancilla_bitsize > 0:
```
Comment (anurudhp) [inline qualtran/bloqs/block_encoding/product.py:158]: ditto, can use `qualtran.symbolics.is_zero`

File: qualtran/bloqs/block_encoding/product.py (hunk 1 above)
Comment (mpharrigan) [reply qualtran/bloqs/block_encoding/product.py]: it's more of an art than a science; but I'd say skip it!

**Review by mpharrigan — APPROVED** [review]

File: qualtran/bloqs/block_encoding/product.py (hunk 1 above)
Comment (charlesyuan314) [reply qualtran/bloqs/block_encoding/product.py]: Removed it.

**Pull request merged**